modules/seastate/src/SeaSt_WaveField.f90

```diff
@@ -432,6 +423,442 @@ SUBROUTINE WaveField_GetWaveKin( WaveField, SeaSt_Interp_m, Time, pos, forceNode
+   character(*),      intent(  out)  :: ErrMsg
+
+   real(ReKi)                        :: Tmp
+
+   ErrStat = ErrID_None
+   ErrMsg  = ""
+
+   isopc   = -1.0
+   Indx_Lo = 0
+   Indx_Hi = 0
+
+
+   !Tmp =  acos(-p / z_depth) / delta
+   Tmp = acos( max(-1.0_ReKi, min(1.0_ReKi, 1+(p / z_depth)) ) ) / delta
+   Tmp =  nmax - 1 - Tmp
+   Indx_Lo = INT( Tmp ) + 1    ! convert REAL to INTEGER, then add one since our grid indices start at 1, not 0
+   isopc = 2.0_ReKi * (Tmp - REAL(Indx_Lo - 1, ReKi)) - 1.0_ReKi  ! convert to value between -1 and 1
+
+   if ( Indx_Lo < 1 ) then
+      Indx_Lo = 1
+      isopc = -1.0
+      if (FirstWarn) then
+         call SetErrStat(ErrID_Warn,'Position has been clamped to the grid boundary. Warning will not be repeated though condition may persist.',ErrStat,ErrMsg,'SetCartesianZIndex') !error out if z is outside the lower bounds
+         FirstWarn = .false.
+      end if
+   end if
+
+   Indx_Hi = min( Indx_Lo + 1, nMax )     ! make sure it's a valid index, one-based
+
+   if ( Indx_Lo >= Indx_Hi ) then
+      ! Need to clamp to grid boundary
+      if (FirstWarn .and. Indx_Lo /= Indx_Hi) then ! don't warn if we are exactly at the boundary
+         call SetErrStat(ErrID_Warn,'Position has been clamped to the grid boundary. Warning will not be repeated though condition may persist.',ErrStat,ErrMsg,'SetCartesianZIndex') !error out if z is outside the upper bounds
+         FirstWarn = .false.
+      end if
+      Indx_Lo = max(Indx_Hi - 1, 1)
+      isopc = 1.0
+   end if
+
+   !-------------------------------------------------------------------------------------------------
+   ! to verify that we don't extrapolate, make sure isopc is bound between -1 and 1 (effectively nearest neighbor)
+   !-------------------------------------------------------------------------------------------------
+   isopc = min( 1.0_SiKi, isopc )
+   isopc = max(-1.0_SiKi, isopc )
+
+end subroutine SetCartesianZIndex
+
+
+subroutine SetTimeIndex(Time, deltaT, nMax, Indx_Lo, Indx_Hi, isopc, ErrStat, ErrMsg)
+   real(DbKi),        intent(in   )  :: Time     !< time from the start of the simulation
+   real(ReKi),        intent(in   )  :: deltaT
+   integer(IntKi),    intent(in   )  :: nMax
+   integer(IntKi),    intent(inout)  :: Indx_Lo
+   integer(IntKi),    intent(inout)  :: Indx_Hi
+   real(SiKi),        intent(inout)  :: isopc
+   integer(IntKi),    intent(  out)  :: ErrStat
+   character(*),      intent(  out)  :: ErrMsg
+
+   real(ReKi)                        :: Tmp
+
+   ErrStat = ErrID_None
+   ErrMsg  = ""
+
+   isopc   = -1.0
+   Indx_Lo = 0
+   Indx_Hi = 0
+   if ( Time < 0.0_DbKi ) then
+      CALL SetErrStat(ErrID_Fatal,'Time value must be greater than or equal to zero!',ErrStat,ErrMsg,'SetTimeIndex') !error out if time is outside the lower bounds
+      RETURN
+   end if
+
+   ! if there are no timesteps, don't proceed
+   if (EqualRealNos(deltaT,0.0_ReKi) .or. deltaT < 0.0_ReKi)  return;
```

Comment:
@luwang00 & @RBergua, this check was the root cause of the seg-faults.

Reply:
@andrew-platt Thanks for fixing this!